src/models/geograph.py

```diff
@@ -614,21 +669,317 @@ def get_graph_components(
         https://en.wikipedia.org/wiki/Component_(graph_theory)
 
         Args:
-            graph (nx.Graph): nx.Graph object. This must be a subgraph of this
-            class's main `graph` attribute.
+            calc_polygons (bool, optional): This determines whether to calculate
+            the polygons which are the union of all the polygons that make up
+            each component, and load a `ComponentGeoGraph` with a corresponding
+            dataframe containing these components. This can be time consuming
+            if there is a very large component. Defaults to True.
+            add_distance_edges (bool, optional): This determines whether to add
+            edges between every pair of nodes, with the distance between their
+            corresponding polygons as an edge attribute. Defaults to False.
 
         Returns:
-            Tuple: A tuple containing the resulting GeoDataFrame and the list of
-            graph components.
+            ComponentGeoGraph: A ComponentGeoGraph containing the resulting
+            GeoDataFrame (if `calc_polygons=True`) and the list of graph
+            components.
         """
-        components: List[set] = list(nx.connected_components(graph))
-        geom = [self.df.geometry.loc[comp].unary_union for comp in components]
-        gdf = gpd.GeoDataFrame({"geometry": geom}, crs=self.df.crs)
-        return gdf, components
+        components: List[set] = list(nx.connected_components(self.graph))
+        if calc_polygons:
+            geom = [self.df["geometry"].loc[comp].unary_union for comp in components]
+            gdf = gpd.GeoDataFrame(
+                {"geometry": geom, "class_label": -1}, crs=self.df.crs
+            )
+            comp_geograph = ComponentGeoGraph(
+                components, gdf, add_distance_edges=add_distance_edges
+            )
+        else:
+            comp_geograph = ComponentGeoGraph(components)
+        if not hasattr(self, "components"):
+            self.components = comp_geograph
+        return comp_geograph
+
+    def get_metric(self, name: str) -> metrics.Metric:
```

Comment:
Just for my understanding: so if I want to get all available metrics I could just use the line below?
```python
all_metrics_of_graph = [graph.get_metric(name) for name in metrics.METRICS_DICT]
```

Reply:
The problem is that `avg_component_isolation` doesn't work unless you have edges with distance between the components in the `ComponentsGeoGraph` of the `GeoGraph` you call the metric on, and `avg_component_area` doesn't work when the `ComponentsGeoGraph` doesn't have a dataframe with polygons, which is the default for the main `GeoGraph`. So if you do the above then you will get an error - any ideas on how to resolve this?

Reply:
Maybe a property like `available_metrics` or so would be nice? This way a user could check what possible arguments are allowed.

Reply:
One thing you can do is 
```python
for name in metrics.METRICS_DICT:
    try:
        all_metrics.append(graph.get_metric(name))
    except ValueError:
        continue
```
which would not give any errors and should return all the metrics it is possible to calculate.

Reply:
I think given the computational cost to create some of these metrics scales so poorly we should keep the GeoGraph defaults scalable (i.e. not compute all the component edges, df etc.). The most important immidiate thing will be to add good warnings (and docstrings with examples) that let the user know how to add the relevant info to compute a more sophisticated metric.

In the next step, we could let the metrics themselves trigger the computation of these data structures. So e.g. if you want to compute `avg_component_area` it calls a `GeoGraph` method `get_components(datafram=True)`, that computes the components with df if the current component attribute doesn't have this yet. This would make it easier for the user, but maintain default scalability.

Reply:
We can certainly make it so that the metrics functions trigger the computation of the needed data structures, but that would then make it impractical to use the code above iterating over all metrics, because it would trigger some incredibly expensive computation unless your graph is small. I personally slightly prefer allowing users to get all metrics calculatable with the defaults easily. Maybe I should add a `get_all_metrics` function which gets all the metrics that are valid for a geograph - then additional ones can be called optionally.

Reply:
Hm yeah I think you're right, it would be nice for the user to have an easy way to get the simple metrics. Maybe we could have `get_standard_metrics` (for computationally inexpensive) and `get_extended_metrics` (for all metrics including expensive) functions?

Reply:
Generally, I think triggering these computations is what a user would expect, if you try to get a complicated metric. But you're absolutely right, it shouldn't be the default anywhere.

Reply:
Alright, in b196db1 I added a `STANDARD_METRICS` list and an `ALL_METRICS` list, so you can just iterate over the standard metrics list to get all of the cheap metrics. Then I made the metrics functions compute the needed data structures automatically, with a warning.